File: python_75/two/prod_of_arr_ex_self.py

# input = [1, 2, 3, 4]
# output = [24, 12, 8, 6]

nums = [1, 2, 3, 4]

pre = 1
post = 1

result = [1]*4
print(result)

i = 1
# result = [1, 1, 2, 6]
while(i<len(nums)):
  result[i] = pre 
  pre = pre * nums[i]
  i += 1

i = len(nums) - 1
while (i>=0):
  result[i] = post * result[i]
  post = post * nums[i]
  i -= 1
print(result)

File: spring-data-redis-example/src/main/java/com/sunil/redis/ds/prodOfArrExSelf.java
package com.sunil.redis.ds;

import java.util.Arrays;

public class prodOfArrExSelf {

    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 4};
        System.out.println(Arrays.toString(prod(nums)));
    }

    public static int[] prod(int[] nums) {
        int pre = 1;
        int post = 1;

        int[] result = new int[nums.length];

        for (int i=0; i<nums.length; i++) {
            result[i] = pre;
            pre = pre * nums[i];
        }
//        [1, 1, 2, 6]
        for (int i=nums.length -1; i>=0; i--) {
            result[i] = result[i] * post;
            post = post * nums[i];
        }

        return result;
    }
}
